src/silvimetric/resources/metric.py

```diff
@@ -64,89 +69,136 @@ def __eq__(self, other):
     return np.var(data)
 
-def m_cv(data):
+def m_cv(data, htthreshold, coverthreshold):
     return np.std(data) / np.mean(data)
 
 # TODO check performance of other methods
-def m_abovemean(data):
+def m_abovemean(data, htthreshold, coverthreshold):
     return (data > np.mean(data)).sum() / len(data)
 
 # TODO check performance of other methods
-def m_abovemode(data):
-    return (data > stats.mode(data).mode).sum() / len(data)
+def m_abovemode(data, htthreshold, coverthreshold):
+    return (data > m_mode(data, htthreshold, coverthreshold)).sum() / len(data)
+
+# both stats.skew and stats.kurtosis throw a warning:
+# Precision loss occurred in moment calculation due to catastrophic cancellation. 
+# This occurs when the data are nearly identical. Results may be unreliable.
+#
+# I think this is because values are very similar...end up close to the mean
+# GridMetrics computes these using a different formula that may also have trouble
+# with numeric stability.
+def m_skewness(data, htthreshold, coverthreshold):
+    if len(data) < 4:
+        return -9999.0
+
+    try:
+        s = stats.skew(data)
+    except:
+        s = -9999.0
 
-def m_skewness(data):
-    return stats.skew(data)
+    return s
+
+def m_kurtosis(data, htthreshold, coverthreshold):
+    if len(data) < 4:
+        return -9999.0
 
-def m_kurtosis(data):
-    return stats.kurtosis(data)
+    try:
+        k = stats.kurtosis(data)
+    except:
+        k = -9999.0
 
-def m_aad(data):
-    m = np.mean(data)
+    return k
+
+def m_aad(data, htthreshold, coverthreshold):
+    m = m_mean(data, htthreshold, coverthreshold)
     return np.mean(np.absolute(data - m))
 
-def m_madmedian(data):
+def m_madmedian(data, htthreshold, coverthreshold):
     return stats.median_abs_deviation(data)
 
-def m_madmean(data):
+def m_madmean(data, htthreshold, coverthreshold):
     return stats.median_abs_deviation(data, center=np.mean)
 
-def m_madmode(data):
-    stats_mode = lambda v,axis : stats.mode(v,axis).mode
-    return stats.median_abs_deviation(data, center=stats_mode)
+# TODO needs work
+def m_madmode(data, htthreshold, coverthreshold):
+    m = m_mode(data, htthreshold, coverthreshold)
+    return np.median(np.absolute(data - m))
 
 # TODO test various methods for interpolation=... for all percentile-related metrics
 # I think the default matches FUSION method but need to test
-def m_iq(data):
+def m_iq(data, htthreshold, coverthreshold):
     return stats.iqr(data)
 
-def m_90m10(data):
+def m_90m10(data, htthreshold, coverthreshold):
     p = np.percentile(data, [10,90])
     return p[1] - p[0]
 
-def m_95m05(data):
+def m_95m05(data, htthreshold, coverthreshold):
     p = np.percentile(data, [5,95])
     return p[1] - p[0]
 
-def m_crr(data):
-    return (np.mean(data) - np.min(data)) / (np.max(data) - np.min(data))
+def m_crr(data, htthreshold, coverthreshold):
+    maxv = np.max(data)
+    minv = np.min(data)
+    if minv == maxv:
+        return -9999.0
```

Comment:
Agreed...This could be a constant or part of config and, potentially, specified on CL. Problem with making it part of config is that it would have to be passed to the metric functions. I suppose the metric functions could return None and then we deal with assigning NODATA where we have access to config info